From 54ba3bc8a55c50e4712997e0d7867f23498d0597 Mon Sep 17 00:00:00 2001
From: Timo Herbrecher <t.herbrecher@gateware.de>
Date: Thu, 4 Apr 2019 16:59:25 +0200
Subject: [PATCH 1/1] mtd/fsl-quadspi: Fix quadspi read error after ioremap

A pointer to an ioremapped memory should not be treated as regular
memory pointer regarding to the kernel API documentation. Therefore
memcpy should not be used for reading data from the QuadSPI memory
region.
This causes segmentation faults (DABT) in some conditions, e.g. mounting
ubifs filesystems.
Replace 'memcpy' with the io-save version 'memcpy_fromio'.

Signed-off-by: Timo Herbrecher <t.herbrecher@gateware.de>
---
 drivers/mtd/spi-nor/fsl-quadspi.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/drivers/mtd/spi-nor/fsl-quadspi.c b/drivers/mtd/spi-nor/fsl-quadspi.c
index 9e95a9054bd3..d5f0eb9277de 100644
--- a/drivers/mtd/spi-nor/fsl-quadspi.c
+++ b/drivers/mtd/spi-nor/fsl-quadspi.c
@@ -926,7 +926,7 @@ static ssize_t fsl_qspi_read(struct spi_nor *nor, loff_t from,
 		len);
 
 	/* Read out the data directly from the AHB buffer.*/
-	memcpy(buf, q->ahb_addr + q->chip_base_addr + from - q->memmap_offs,
+	memcpy_fromio(buf, q->ahb_addr + q->chip_base_addr + from - q->memmap_offs,
 		len);
 
 	return len;
-- 
2.17.1

